modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/syncfs/SyncFsUtils.java

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright 2024 GridGain Systems, Inc. and Contributors.
+ *
+ * Licensed under the GridGain Community Edition License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.gridgain.com/products/software/community-edition/gridgain-community-edition-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.cache.persistence.checkpoint.syncfs;
+
+import com.sun.jna.Native;
+import java.io.IOException;
+import java.nio.file.FileStore;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentMap;
+import org.apache.ignite.IgniteCheckedException;
+import org.apache.ignite.internal.pagemem.store.PageStore;
+import org.apache.ignite.internal.processors.cache.persistence.StorageException;
+import org.apache.ignite.internal.processors.cache.persistence.checkpoint.CheckpointPagesWriter.CheckpointPageStoreInfo;
+import org.apache.ignite.internal.processors.cache.persistence.file.FilePageStore;
+import org.apache.ignite.internal.util.collection.IntHashMap;
+import org.apache.ignite.lang.IgniteBiTuple;
+
+import static org.apache.ignite.IgniteSystemProperties.getBoolean;
+import static org.apache.ignite.IgniteSystemProperties.getInteger;
+
+/**
+ * Class that holds utilities for "syncfs" function that can be used during checkpoint.
+ */
+public class SyncFsUtils {
+    /**
+     * Name of the system property that enables {@code syncfs} checkpoint optimization.
+     */
+    public static final String IGNITE_CHECKPOINT_SYNC_FS_ENABLED = "IGNITE_CHECKPOINT_SYNC_FS_ENABLED";
+
+    /**
+     * Name of the system property that controls the threshold for the number of files that will trigger the
+     * {@code syncfs} instead of multiple {@code fsync} calls for each individual file in the checkpoint.
+     */
+    public static final String IGNITE_CHECKPOINT_SYNC_FS_THRESHOLD = "IGNITE_CHECKPOINT_SYNC_FS_THRESHOLD";
+
+    /**
+     * Value of the {@link #IGNITE_CHECKPOINT_SYNC_FS_ENABLED} system property. {@code true} by default.
+     */
+    public static final boolean SYNC_FS_ENABLED = getBoolean(IGNITE_CHECKPOINT_SYNC_FS_ENABLED, true);
+
+    /**
+     * Value of the {@link #IGNITE_CHECKPOINT_SYNC_FS_THRESHOLD} system property. {@code 1024} by default.
+     */
+    public static final int SYNC_FS_THRESHOLD = getInteger(IGNITE_CHECKPOINT_SYNC_FS_THRESHOLD, 1024);
```

Comment:
How is this 1024 obtained?

Reply:
Also, is this 1024 files per checkpoint, or 1024 files per filesystem per checkpoint? Could it happen that just 2 files need to be fsynced on one FS and 10k files on another one? If yes, will this cause FS sync on both filesystems?

Reply:
1024 per checkpoint, determining how many FSs there are requires some extra code as well. This number is almost random, I must say, because there's no way I can predict it precisely.
I see what you mean, but I'm not sure that such cases worth the effort